core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -507,83 +507,92 @@ class KafkaApis(val requestChannel: RequestChannel,
       // greater than that supported by the client (as indicated by the fetch request version). If the
       // configured magic version for the topic is less than or equal to that supported by the version of the
       // fetch request, we skip the iteration through the records in order to check the magic version since we
       // know it must be supported. However, if the magic version is changed from a higher version back to a
       // lower version, this check will no longer be valid and we will fail to down-convert the messages
       // which were written in the new format prior to the version downgrade.
-      replicaManager.getMagic(tp).flatMap { magic =>
-        val downConvertMagic = {
-          if (magic > RecordBatch.MAGIC_VALUE_V0 && versionId <= 1 && !unconvertedRecords.hasCompatibleMagic(RecordBatch.MAGIC_VALUE_V0))
-            Some(RecordBatch.MAGIC_VALUE_V0)
-          else if (magic > RecordBatch.MAGIC_VALUE_V1 && versionId <= 3 && !unconvertedRecords.hasCompatibleMagic(RecordBatch.MAGIC_VALUE_V1))
-            Some(RecordBatch.MAGIC_VALUE_V1)
-          else
-            None
-        }
-
-        downConvertMagic.map { magic =>
-          trace(s"Down converting records from partition $tp to message format version $magic for fetch request from $clientId")
+      val records = partitionData.records
+      val downConvertMagic =
+        replicaManager.getMagic(tp).flatMap { magic =>
+            if (magic > RecordBatch.MAGIC_VALUE_V0 && versionId <= 1 && !records.hasCompatibleMagic(RecordBatch.MAGIC_VALUE_V0))
+              Some(RecordBatch.MAGIC_VALUE_V0)
+            else if (magic > RecordBatch.MAGIC_VALUE_V1 && versionId <= 3 && !records.hasCompatibleMagic(RecordBatch.MAGIC_VALUE_V1))
+              Some(RecordBatch.MAGIC_VALUE_V1)
+            else
+              None
+          }
 
-          // Because down-conversion is extremely memory intensive, we want to try and delay the down-conversion as much
-          // as possible. With KIP-283, we have the ability to lazily down-convert in a chunked manner. The lazy, chunked
-          // down-conversion always guarantees that at least one batch of messages is down-converted and sent out to the
-          // client.
-          new LazyDownConversionRecords(tp, unconvertedRecords, magic, fetchContext.getFetchOffset(tp).get, time)
-        }
-      }.getOrElse(unconvertedRecords)
+      // For fetch requests from clients, check if down-conversion is disabled for the particular partition
+      if (!fetchRequest.isFromFollower && downConvertMagic.isDefined && !replicaManager.messageDownConversionEnabled(tp).getOrElse(true)) {
+        trace(s"Conversion to message format ${downConvertMagic.get} is disabled for partition $tp. Sending unsupported version response to $clientId.")
+        new FetchResponse.PartitionData(Errors.UNSUPPORTED_VERSION, FetchResponse.INVALID_HIGHWATERMARK,
```

Comment:
Could we use `errorResponse`?

Reply:
Unfortunately, no. `errorResponse` returns `FetchResponse.PartitionData[Records]` but we want `FetchResponse.PartitionData[BaseRecords]` here.

Reply:
Found the incantation to make this work:
```scala
    def errorResponse[T >: MemoryRecords <: BaseRecords](error: Errors): FetchResponse.PartitionData[T] = {
      new FetchResponse.PartitionData[T](error, FetchResponse.INVALID_HIGHWATERMARK, FetchResponse.INVALID_LAST_STABLE_OFFSET,
        FetchResponse.INVALID_LOG_START_OFFSET, null, MemoryRecords.EMPTY)
    }
```
This says that `T` is a supertype of `MemoryRecords` and a subtype of `BaseRecords`. Since `Records` and `BaseRecords` both satisfy this bound, it works.